### PR TITLE
sdk: Add support for authenticated media requests

### PR DESCRIPTION
**Pull request by zecakeh**
This follows the [implementation guidance](https://matrix.org/blog/2024/06/20/matrix-v1.11-release/#client-implementation-guidance) to only use the authenticated endpoints when the server advertises support for Matrix 1.11.

- [x] Public API changes documented in changelogs (optional)

### CODE REVIEW

File: crates/matrix-sdk/src/media.rs [hunk 1]
```diff
@@ -269,10 +272,19 @@ impl Media {
             }
         };
 
+        let server_versions = self.client.server_versions().await?;
```
Comment (poljar) [inline crates/matrix-sdk/src/media.rs:275]: This allows the homeserver to downgrade us towards unauthenticated media uploads. While it isn't the end of the world, I think we should provide a `ClientBuilder` option that allows users to always use the authenticated endpoints, despite what the homeserver suggests.

This doesn't need to happen as part of this PR, but please leave a TODO item here.

**Review by poljar — CHANGES_REQUESTED** [review]
Left a small nit, please get rid of the merge conflicts and add the TODO item and we can merge.

File: crates/matrix-sdk/src/media.rs (hunk 1 above)
Comment (zecakeh) [reply crates/matrix-sdk/src/media.rs]: Todo added

**Comment by zecakeh** [comment]
I merged main and added the todo (with a small refactoring).

**Review by poljar — APPROVED** [review]
Looks good. Thanks.

**Pull request merged**